src/reflect/scala/reflect/internal/tpe/GlbLubs.scala

```diff
@@ -230,14 +230,17 @@ private[internal] trait GlbLubs {
   def weakLub(tps: List[Type]): Type = (
     if (tps.isEmpty)
       NothingTpe
-    else if (tps forall isNumericValueType)
+    else if (shouldUseNumericLub(tps))
       numericLub(tps)
     else if (tps exists typeHasAnnotations)
       annotationsLub(lub(tps map (_.withoutAnnotations)), tps)
     else
       lub(tps)
   )
 
+  private def shouldUseNumericLub(tps: List[Type]) =
+    tps forall (tp => isNumericValueType(tp.widen))
```

Comment:
`.deconst` is perfectly decent here, I think; I tried pretty hard to find an observable difference, but since [single types are evidently expected to conform to `AnyRef`](https://www.scala-lang.org/files/archive/spec/2.12/03-types.html#singleton-types) I can't trip up the lubber with something like 

```scala
  val i1 = 1
  val l1 = 1L

  type I1 = i1.type
  type L1 = l1.type

  def foo3(a: Option[Any]) = (a : @unchecked) match { case Some(x @ (I1 | L1)) => x }
  foo3(None) : Long
```

... which is probably for the best.

As I mentioned, I mostly but still don't quite understand what the other use of `weakLub` (being `isWeakSubType` in `TypeComparers`) uses it for, but its arguments [are already `dealias`ed](https://github.com/scala/scala/blob/1b7d57f2f90b2c851a5c60250c077721d01c48f8/src/reflect/scala/reflect/internal/tpe/TypeComparers.scala#L571) by the time they get there, so that shouldn't be a worry.

The fact that all of the (par)tests pass with `isNumericValueType` doing the `widen`ing, even in the singleton literals PR, gives me great hope that it's not the wrong thing to do. I'm more than happy to move it there to give 2.12.4 a sneak peek of the greatness which that PR will bring.

Does that sound reasonable to you?